### PR TITLE
[GEOS-11294] MapML Add Support to WMS for Vector Representation

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-11294](https://osgeo-org.atlassian.net/browse/GEOS-11294) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

useFeatures UI test

added usefeatures to builder

Return features links

started on conversion

First pass at building features

clarify multi error

cleanup

PR comments part 1

Configuration Panel Constants

check for raster and more constants

fixed metadata tests

added getAsMapML to tests

removed alternative projection links and added feature reprojection

test issue

more test cleanup

fixed default num decimals

reprojectingfeaturecollection needs fallback CRS

if feature links, no query links

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]
Need a bit of documentation for the new GUI option.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java [hunk 1]
```diff
@@ -31,6 +31,69 @@ public final class MapMLConstants {
     /** format name */
     public static final String FORMAT_NAME = "MAPML";
 
+    /** MapML format options */
+    public static final String MAPML_FEATURE_FORMAT_OPTIONS = "mapmlfeatures:true";
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLConstants.java:35]: It's all or nothing, the format option doesn't show up at all if features are not requested. So, mapmlfeatures:false never happens.

**Comment by turingtestfail** [comment]
> Need a bit of documentation for the new GUI option.

Added

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java [hunk 2]
```diff
@@ -348,6 +354,7 @@ private MapMLLayerMetadata layersToOneMapMLLayerMetadata(List<RawLayer> layers)
         MapMLLayerMetadata mapMLLayerMetadata = new MapMLLayerMetadata();
         mapMLLayerMetadata.setLayerMeta(new MetadataMap());
         mapMLLayerMetadata.setUseTiles(false);
+        useFeaturesAllLayers = isAllLayersUseFeatures(layers);
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java]: Done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java [hunk 3]
```diff
@@ -362,6 +369,24 @@ private MapMLLayerMetadata layersToOneMapMLLayerMetadata(List<RawLayer> layers)
         return mapMLLayerMetadata;
     }
 
+    /**
+     * Check if all layers in the request use features
+     *
+     * @param layers List of RawLayer objects
+     * @return boolean
+     */
+    private boolean isAllLayersUseFeatures(List<RawLayer> layers) {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java]: Done

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java [hunk 4]
```diff
@@ -536,7 +564,13 @@ private MapMLLayerMetadata layerToMapMLLayerMetadata(RawLayer layer, String styl
                         && gwc.getTileLayer(isLayerGroup ? layerGroupInfo : layerInfo)
                                         .getGridSubset(projType.value())
                                 != null;
-        boolean useTiles = Boolean.TRUE.equals(layerMeta.get("mapml.useTiles", Boolean.class));
+        boolean useTiles = Boolean.TRUE.equals(layerMeta.get(MAPML_USE_TILES, Boolean.class));
+        boolean useFeatures = Boolean.TRUE.equals(layerMeta.get(MAPML_USE_FEATURES, Boolean.class));
+        if (!useFeatures || !isVector) {
+            // if any layer in the request is not using features or is not vector type,
+            // then the request as a whole is not using features
+            useFeaturesAllLayers = false;
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java]: D'oh.  Not my finest work.  Fixed.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeaturesBuilder.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        SettingsInfo settings = geoServer.getSettings();
+        if (meta instanceof FeatureTypeInfo) {
+            FeatureTypeInfo featureTypeInfo = (FeatureTypeInfo) meta;
+            if (featureTypeInfo.getNumDecimals() > 0) {
+                return featureTypeInfo.getNumDecimals();
+            }
+            return settings.getNumDecimals();
+        }
+
+        return settings.getNumDecimals();
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLFeaturesBuilder.java]: Done

**Comment by aaime** [comment]
The doc failure is unrelated, I've warned Jody in the PR that introduced the issue.

**Pull request merged**